### ISSUE TITLE
[v6] FieldArray Radio Example

### ISSUE THREAD

**Issue by tyrsius** [erikras/redux-form#1226]
I've looked at these [two](https://github.com/erikras/redux-form/issues/124) [issues](https://github.com/erikras/redux-form/issues/33), but they are either for v5 or they do not actually have a working example.

Right now the behavior of radio buttons is either strange or broken. The UI updates the way it should, but the value in `onSubmit` is `on` for every field that was ever checked as a radio button. I don't see a way to tell in the form submission event to tell which item in `FieldArray` had its radio button checked.

**Comment by erikras** [comment]
A radio group is a single field. See `sex` on the [Simple Example](http://redux-form.com/6.0.0-alpha.15/examples/simple/).

**Comment by tyrsius** [comment]
That isn't in a `FieldArray`. the `value` of a Radio in a `FieldArray` is relative to the item, whose value you don't have access to.

**Comment by tnrich** [comment]
Hey @tyrsius , my guess is that you're using the FieldArray incorrectly. Maybe try double checking that your use looks like this example: http://redux-form.com/6.0.0-alpha.15/examples/fieldArrays/

**Comment by tnrich** [comment]
Note how just using a <FieldArray> alone isn't enough, you also need a <Field> component to handle the actual toggling of the radio button:

```js
const renderMembers = ({ fields }) => (
  <ul>
    <li>
      <button type="button" onClick={() => fields.push({})}>Add Member</button>
    </li>
    {fields.map((member, index) =>
      <li key={index}>
        <button
          type="button"
          title="Remove Member"
          onClick={() => fields.remove(index)}/>
        <h4>Member #{index + 1}</h4>
        <Field
          name={`${member}.firstName`}
          type="text"
          component={renderField}
          placeholder="First Name"/>
```

**Comment by tyrsius** [comment]
@tnrich perhaps an example with a radio button would be more helpful, as my problem is not understanding how to set the `checked` property.

**Comment by babsonmatt** [comment]
Does this mean an error on the field "sex" would show for both radio buttons? (Assuming each was a custom component with an error display method) If so, how would you prevent that and just show an error for the group?

**Comment by tyrsius** [comment]
@babsonmatt I haven't been able to get the value binding to work, I haven't gotten as far as showing errors.

**Comment by tnrich** [comment]
@babsonmatt 

This is a cobbled together example from my app of an array of radio buttons controlled by a FieldArray and individual Field components  (it's using a custom radio component and is a bit verbose but gets the point across):

If you want an array of radio inputs, the key point is the same as I tried to explain above: 
```js
//in some render method:
        <FieldArray 
          name="inserts"
          component={renderMultipleRadios}/> 

//function defined outside of render method
const renderMultipleRadios = ({ fields }) => (
  <ul>
    <li>
      <button type="button" onClick={() => fields.push({})}>Add Radio</button>
    </li>
    {fields.map((radio, index) =>
      <li key={index}>
        <Field
          name={`${radio}.firstName`}
          component={renderRadio} //this is the same 
          />




//another function defined outside of render:
function renderRadio (input) {
    return <div className="radioHolder">
      <Radio
        className="RestrictionDigestRadio"
        onChange={() => { 
          input.onChange('RestrictionDigest') }}
          checked={input.value === 'RestrictionDigest'}
          label='Restriction Digest'
          />
        <Radio
          onChange={() => { 
            input.onChange('PCR') }}
            checked={input.value === 'PCR'}
            label='PCR'
            />
      </div>
}

```

I hope that makes sense for you. Sorry if there are some little errors in the fine print.

**Comment by tyrsius** [comment]
That makes sense, and I can see how that would be useful, but that is a radio value owned within a single item. My use case is different, and its starting to look like it isn't supported.

```
Name | Default
---------------------------
1    | O
2    | O
3    | X (filled in radio)
```

I have an array of children, and a Radio that is part of a group across the children, which tells the top level form which is the "default" one. This means the Radio isn't owned by a single `<Field>`, and the values aren't hard-coded.

I will have to find another solution

**Comment by tnrich** [comment]
@tyrsius I'm not 100% sure I understand your use case, but it seems like it can still be solved using redux-form v6. I would maybe try passing an extra prop to the Field:

```js
{fields.map((member, index) =>
      <li key={index}>
        <button
          type="button"
          title="Remove Member"
          onClick={() => fields.remove(index)}/>
        <h4>Member #{index + 1}</h4>
        <Field
          name={`${member}.firstName`}
          type="text"
          handleDefaultChecked={function(indexOfNewDefault){ setNewDefault(indexOfNewDefault) }}
          component={renderField}
          placeholder="First Name"/>
```

You could use vanilla redux to create that "setNewDefault" action or you could use another redux field higher up in the rendering stack! 

I'm not 100% sure, but I think that should do what you're wanting.

**Comment by tyrsius** [comment]
The way I'm solving it now is kind of a mess, I'll try to trim out the irrelevant code.

```jsx
const ApiVersionEdit = (props) => (
  <form onSubmit={props.handleSubmit((data) => {
    let newApi = {...data}
    let latestVersion = newApi.versions[props.latestVersionIndex]
    newApi.latestVersion = latestVersion && latestVersion.versionName
    return props.onSubmit(newApi)
  })}>
    <FieldArray name="versions" component={renderVersions}
      props={{
        latestVersionIndex: props.latestVersionIndex,
        updateLatest: props.updateLatest
      }} />

    <SaveButton {...props} label="Save" />
  </form>
)

const renderVersions = ({fields, statuses, newVersionMap, latestVersionIndex, updateLatest}) => (
  <div>
    {fields.map((version, index) => (
      <tr key={index}>
        <td>
          <Field name={`${version}.isLatest`}
            radioName="isLatest"
            checked={index === latestVersionIndex}
            onChange={(e) => {
              updateLatest(index)
            }}
            component={Radio} />
        </td>
      </tr>
    ))
  }
  </div>
)

let ApiVersionEditForm = reduxForm({
  form: 'apiVersionEdit',
  validate
})(ApiVersionEdit)

const selector = formValueSelector('apiVersionEdit')
ApiVersionEditForm = connect(state => {
  let versions = selector(state, 'versions') || []
  let latestVersionIndex = selector(state, 'latestVersionIndex')

  if (latestVersionIndex === undefined) {
    latestVersionIndex = versions.findIndex(v => v.versionName === selector(state, 'latestVersion'))
  }
  return {
    latestVersionIndex,
    newVersionMap: versions.map(v => v.isNew || false)
  }
}, { updateLatest: change.bind(null, 'apiVersionEdit', 'latestVersionIndex') })(ApiVersionEditForm)
```

**Comment by tyrsius** [comment]
My issue with this is that determining the checked status requires a `formValueSelector` over the entire parent field, which is causes a lot of unnecessary re-renders. But I don't know if another way is possible

**Comment by oyeanuj** [comment]
@tnrich @erikras A question that maybe you guys could help answer (since am running into the same issue). In the example posted by @erikras (Simple example: `sex` field) - `FieldArray` isn't necessary. 

So when is it recommended or appropriate to use the `FieldArray` when dealing with a radio button with a few options?

**Comment by vincentaudebert** [comment]
Agree it would be very useful to add an example with a group of radio buttons in the documentation. It's not that clear how to do it.

**Comment by vincentaudebert** [comment]
Might help someone: https://gist.github.com/vincentaudebert/2aba73ec8f10cd141bafe21f660bf7c8

**Comment by grovesNL** [comment]
@vincentaudebert Your example doesn't use `FieldArray` though

@tyrsius Did you figure out how to do this in a cleaner way? I have the same issue and was considering an approach like yours

**Comment by tyrsius** [comment]
@grovesNL There is no good way. I am actually using redux-form for everything else, and using a standard radio button element with an `onChange` event that gets stored and merged with the form result later

**Comment by grovesNL** [comment]
@tyrsius Thanks, that's too bad. I'll probably do the same.

**Comment by danielrob** [comment]
@tyrsius it took me three reads of this thread, but I think I get what you're after now... Yes, ultimately you're going to need to do some custom connecting in this situation. I would recommend something like this (psuedo code): 

```
renderFieldArray({ fields }) => (
  fields.map((indexedFieldName, index) =>  (   
    <div>
       <Field name="latestVersion" component="input" type="radio" value={index} />
       <FieldGroup name={indexedFieldName} />
    <div>
  )
)

const selector = formValueSelector('myFormName')

const FieldGroup = ({ name, latestVersion }) => (
  <div>
      // Render a group of Fields/Fields/FieldArrays/Whatevers whose behaviour changes according to whether they are the latestVersion or not. 
  </div>
)

export default connect(
  state => selector(state, 'latestVersion')
)(FieldGroup)
``` 

This will be more efficient since <FieldGroup> should not re-render unless it's props change. (General rule when connecting components is as low and late as possible when it comes to performance). 

**tl;dr**
1. Essentially, a team of radio buttons should work on one value in the store => namely `latestVersion`. Then, subscribe to that value at the latest possible moment where it needs to be consumed. 
2. A group of radio buttons not operating on the same key are not a group of radio buttons; they're a group of size-1 radio button groups.

**Comment by grovesNL** [comment]
@danielrob The point is that redux-form is supposed to manage those concerns, otherwise I get very little benefit from attaching redux-form to those buttons. I can work around it and wire up many custom actions, but it is far too easy to get the lifecycle wrong when dealing with nested components.

Currently I use what I consider a "safe" subset of the redux-form API ,which includes each of the components (i.e. `<Field />`) and very few of the selector/action creators, with the hope that redux-form will eventually support this kind of capability. I've been considering writing my own abstraction over redux-form to avoid repeating these workarounds everywhere, or switching to another library.

**Comment by danielrob** [comment]
@grovesNL it's just a bit unclear what concerns you're hoping it to manage? As mentioned it took me three reads of this entire thread to have a guess at the use case - not sure if I got it? The one concern I do see reappearing is that field disabling/interdependency could be made more comprehensible, but there is a clear way of handling this, which is with the provides selectors. If you can think of a better way to do this, and what use cases it would achieve, feel free to let it be known!

**Comment by grovesNL** [comment]
@danielrob There are lots of common use cases, for example https://github.com/erikras/redux-form/issues/124#issuecomment-212199756 (the provided solution is a workaround).

If I recall correctly, my original issue is analogous to the following example:

- a user selects an animal from a vertical list of animals (cat, dog, mouse) displayed as radio buttons
- choosing *cat* causes some cat-specific options to be displayed, choosing *dog* displays some dog-specific options, etc.

To implement this with redux-form, somebody might reasonably try to use `FieldArray` to render`Animal` components to handle custom rendering for each animal type. However, this person would quickly hit the issue discussed here and the issue linked above. The radio buttons are not connected to one another, and the default value behavior does not work as one might expect.

The greater issue here is that the redux-form model is coupled to the rendered component hierarchy as discussed in #1453 and #1520. When I want to create forms with conditionally rendered fields depending on dirty state within the same model this coupling starts to fail. In my opinion the coupling should be explicit so I can decide how/when my model is bound to fields -- I don't expect redux-form to be able to anticipate how I plan to render my model. I've solved this in some areas by introducing intermediate view models but this tends to add even more confusion.

**Comment by danielrob** [comment]
@grovesNL that was a lot of great reading, but as far as this issue goes I can only conclude that concerns are being mixed. 

A radio button operates on a single field value. You can create multiple radio buttons to operate on that value either by using a multiple `Field` components connected to the same value, or creating a single custom component for one `Field` that renders a number of radio buttons and manages the final value itself. 

FieldArray essentially provides name management when dealing with an array of field values. 

I don't understand to what end you are trying to combine the two concepts.* 

* If it is for conditional field display, e.g. for enabling only one repetition in a repeating list of fields, then what is not appropriate about using the selectors, (or indeed using `Fields`) provided to make your repeater aware of _the_ checkbox group value to activate that value?

**Comment by grovesNL** [comment]
@danielrob Sorry for not being clear. Perhaps I'll need to create a full example repo to illustrate the issue. Let me know if the following example helps.

Consider a store that contains an array of values, like `animals` below:
```
const store = {
    selectedIndex: 1,
    animals: [
        { name: "dog", options: {/* dog-specific options */} },
        { name: "cat", options: {/* mouse-specific options */} },
        { name: "mouse", options: {/* mouse-specific options */} },
    ]
};
```

Users can add/remove their own animals, change any names, update options, etc. within the form. I have an `Animal` component that renders the contents of each row (outputs a `tr`). So I would use `FieldArray` with the `animals` property to render an `Animal` component per element in the array. I want to render a table like this:

|     | Name | Options |
| -- | ------- | ------- |
| ◻ | dog     | -         |
| ◼ | cat       | -         |
| ◻ | mouse | -         |

*Add Animal*
*Remove Animal*

Basically I can't find a straightforward way to render those radio buttons within each row. This is what I meant by my component hierarchy being coupled to my form model.

Others have attempted to move `selectedIndex` into each element of the array, so the model looks like `{ name: "dog", isSelected: false, options: {/* ... */} }`. This way radio buttons can be rendered per animal, but the radio buttons aren't connected to each other. I don't know how to solve that issue without being able to describe relationships within the model.

**Comment by danielrob** [comment]
@grovesNL This is a sandbox of the my earlier comment. Is this what you're after?
 https://codesandbox.io/s/31rOy5Emr

**Comment by danielrob** [comment]
Closing this question. It does not make sense to provide a FieldArray radio group example, because the two concepts are fairly unrelated as discussed. Higher relationships between *the* value of a radio group, and other parts of the form tree can be managed by the developer in custom components, using whatever form of form value selection is most appropriate.

**Comment by grovesNL** [comment]
@danielrob I didn't have a chance to review in detail because I was on vacation.

The issue is that I believe `formValueSelector` and manually passing a field name/value shouldn't be required to express this relatively trivial relationship.

Perhaps I'm searching for a better abstraction to model relationships between fields instead of having to resort to `formValueSelector`? I don't want to cause additional updating/rendering of my components just because this situation forces me to use it.

**Comment by danielrob** [comment]
@grovesNL I 100% agree with you, though I think what you're discussing is a far wider concern than this specific issue. I said some words along the lines of my personal summary of why we're at where we're at with this on another ticket (partially derived from your links above): 

```
Performance is actually (one of) the main reason(s) in my understanding that the Field* components mostly keep to themselves by default, and then the library provides extra selectors for the cases where the developer explicitly needs the payoff of a small performance loss that allows them to fetch the information they want. 
```

In saying that, I totally agree that 'the developer explicitly needing the payoff' is not an uncommon occurrence, and the tracker has a number of tickets to the effect of asking for help with this. I think we need some proposals in this regard. Aiming for an API design that is both useful and general could lead to some fun times :).

**Comment by grovesNL** [comment]
@danielrob Yeah, agreed. Should we open a new issue to discuss this?
